knimin/lib/data_access.py

```diff
@@ -2399,6 +2399,134 @@ def getProjectNames(self):
+
+    def get_plate_map(self, plate_id):
+        """Gets plate map by id
+
+        Parameters
+        ----------
+        plate_id : int
+            Plate id
+
+        Returns
+        -------
+        list of tuple of (int, int, str)
+            Column id, row id, sample id
+        """
+        sql = """SELECT col, row, sample_id FROM pm.plate_sample
+                 WHERE plate_id = %s"""
+        return self._con.execute_fetchall(sql, [plate_id])
+
+    def set_plate_map(self, plate_id, plate_map):
+        """Sets plate map by id
+
+        Parameters
+        ----------
+        plate_id : int
+            Plate id
+        plate_map : list of tuple of (int, int, str)
+            Column id, row id, sample id
+
+        Returns
+        -------
+        tuple (int, int, int)
+            Numbers of samples added / modified / deleted
+
+        Notes
+        -----
+        An empty sample_id represents an empty well, in which case, the record
+        will be deleted if exists.
+        """
+        (nadd, nmod, ndel) = (0, 0, 0)
+        for (col, row, sample_id) in plate_map:
+            sql = """SELECT sample_id FROM pm.plate_sample WHERE plate_id = %s
+                     AND col = %s AND row = %s LIMIT 1"""
+            old_rec = self._con.execute_fetchone(sql, [plate_id, col, row])
+            if old_rec is None:  # Create new record
+                sql_args = [plate_id, col, row, sample_id]
+                sql = """INSERT INTO pm.plate_sample (plate_id, col, row,
+                            sample_id)
+                         VALUES (%s, %s, %s, %s) RETURNING sample_id"""
+                self._con.execute(sql, sql_args)
+                nadd += 1
+            elif sample_id != '':
+                if old_rec[0] != sample_id:  # Modify existing record
+                    sql_args = [sample_id, plate_id, col, row]
+                    sql = """UPDATE pm.plate_sample SET sample_id = %s
+                             WHERE plate_id = %s AND col = %s AND row = %s
+                             RETURNING sample_id"""
+                    self._con.execute(sql, sql_args)
+                    nmod += 1
+            else:  # Delete existing record
+                sql_args = [plate_id, col, row]
+                sql = """DELETE FROM pm.plate_sample WHERE plate_id = %s
+                         AND col = %s AND row = %s RETURNING sample_id"""
+                self._con.execute(sql, sql_args)
+                ndel += 1
+        return (nadd, nmod, ndel)
+
+    def get_plate_type(self, plate_id):
+        """Gets plate type attributes
+        Parameters
+        ----------
+        plate_id : int
+            0: first plate type in the table
+            >0: plate type of an exisiting plate by id
+
+        Returns
+        -------
+        dict
+            Attributes of a plate type, currently including: plate_type_id,
+            name, cols, rows, notes
+        """
+        plate_type_id = 1
+        if plate_id > 0:
```

Comment:
I don't understand why this if is needed? Can you elaborate please? I have the feeling that this can be reduced to a single SQL query:

```python
sql = """SELECT pt.* FROM pm.plate_type pt JOIN pm.plate USING (plate_type_id) WHERE plate_id = %s"""
return self._con..execute_fetchdict(sql, [plate_id])[0]
```